sklearn/discriminant_analysis.py

```diff
@@ -701,22 +904,163 @@ def fit(self, X, y):
+        check_classification_targets(y)
+        self.classes_, y = np.unique(y, return_inverse=True)
+        n_samples, _ = X.shape
+        n_classes = len(self.classes_)
+
+        if n_classes < 2:
+            raise ValueError('The number of classes has to be greater than'
+                             ' one; got %d class' % (n_classes))
+        if n_samples == n_classes:
+            raise ValueError("The number of samples must be more "
+                             "than the number of classes.")
+
+        if self.priors is None:  # estimate priors from sample
+            _, y_t = np.unique(y, return_inverse=True)  # non-negative ints
+            self.priors_ = np.bincount(y_t) / float(len(y))
+        else:
+            self.priors_ = np.asarray(self.priors)
+
+        if (self.priors_ < 0).any():
+            raise ValueError("priors must be non-negative")
+        if not np.isclose(self.priors_.sum(), 1.0):
+            warnings.warn("The priors do not sum to 1. Renormalizing",
+                          UserWarning)
+            self.priors_ = self.priors_ / self.priors_.sum()
+
+        if self.solver == 'svd':
+            if self.covariance_estimator is not None:
+                raise NotImplementedError(
+                    'covariance estimator is not '
+                    'supported with svd solver. Try another solver')
+            self._solve_svd(X, y)
+        elif self.solver == 'lsqr':
+            self._solve_lsqr(X, y, reg_param=self.reg_param,
+                             covariance_estimator=self.covariance_estimator,
+                             standardize=self.standardize)
+        else:
+            raise ValueError("unknown solver {} (valid solvers are 'svd' "
+                             "and 'lsqr').".format(self.solver))
         return self
 
     def _decision_function(self, X):
         check_is_fitted(self)
 
         X = check_array(X)
-        norm2 = []
-        for i in range(len(self.classes_)):
-            R = self.rotations_[i]
-            S = self.scalings_[i]
-            Xm = X - self.means_[i]
-            X2 = np.dot(Xm, R * (S ** (-0.5)))
-            norm2.append(np.sum(X2 ** 2, 1))
-        norm2 = np.array(norm2).T  # shape = [len(X), n_classes]
-        u = np.asarray([np.sum(np.log(s)) for s in self.scalings_])
-        return (-0.5 * (norm2 + u) + np.log(self.priors_))
+
+        if self.solver == "svd":
+            norm2 = []
+            for i in range(len(self.classes_)):
+                R = self.rotations_[i]
+                S = self.scalings_[i]
+                Xm = X - self.means_[i]
+                X2 = np.dot(Xm, R * (S ** (-0.5)))
+                norm2.append(np.sum(X2 ** 2, 1))
+            norm2 = np.array(norm2).T  # shape = [len(X), n_classes]
+            u = np.asarray([np.sum(np.log(s)) for s in self.scalings_])
+            return -0.5 * (norm2 + u) + np.log(self.priors_)
+
+        if self.solver == "lsqr":
```

Comment:
I agree with @rth, let's add the solver in another PR. It will be easier to review